src/compiler/checker.ts

```diff
@@ -4716,19 +4731,33 @@ namespace ts {
             return result;
         }
 
+        function cloneTypePredicate(predicate: TypePredicate, mapper: TypeMapper): ThisTypePredicate | IdentifierTypePredicate {
+            if (predicate.kind === TypePredicateKind.Identifier) {

```

Comment:
I count multiple places (at least three) where you wouldn't need to branch if you did what I propose. In cases where you do branch, the code would be no larger. So, I conclude there is no doubt the code would be smaller if you follow my suggestion. I will leave it at that.